src/core/render.ts

```diff
@@ -1,57 +1,232 @@
+              sceneResolve(item)
+            }
+          })
+
+          scenePromise.then(itemId => {
+            const match  = Render._canvasPool.filter(renObj => renObj['_canvas'] === canvas)
+            if (match.length === 1) {
+              match[0]._rendering = true;
+              iRender.drawToTexture(match[0]['_index'], itemId, 'dupvideoitem').then(res => {
+                resolve(true);
+              })
+            }
+          })
+        } else {
+          reject(Error('Invalid parameters provided. Please check if the canvas/item is correct.'));
+        }
       })
     }
-    // static setRenderFPS(fps: number) {
-    //   return new Promise(resolve => {
-    //     Render.setFPS(fps).then(() => {
-    //       resolve(true)
-    //     })
-    //   })
-    // }
-
-    static testCanvasToUse(canvasIndex, sceneId) {
-      return new Promise(resolve => {
-        iRender.drawToTexture(canvasIndex, sceneId).then(res => {
-          resolve()
-        })
+
+    /**
+     * params: (canvas: HTMLCanvasElement, source: any)
+     *
+     * source parameter can be Source instance, or source id.
+     *
+     */
+    static renderVideoSource(canvas: HTMLCanvasElement, source: any) {
+      // dupsource doesnt work using the source id.
+      return new Promise((resolve, reject) => {
+        let index;
+        if ((source instanceof Source ||  typeof source === 'string')) {
+          let scenePromise = new Promise(sceneResolve => {
+            if (source instanceof Source) {
+              source.getItemList().then(items => {
+                return items[0].getId();
+              }).then(id => {
+                sceneResolve(id);
+              })
+            } else if (typeof source === 'string' && (/^{[A-F0-9\-]*}$/i.test(source))) {
+              Scene.searchSourcesById(source).then(src => {
+                src[0].getItemList().then(items => {
+                  return items[0].getId();
+                }).then(id => {
+                  sceneResolve(id);
+                })
+              })
+            }
+          })
+
+          scenePromise.then(itemId => {
+            const match  = Render._canvasPool.filter(renObj => renObj['_canvas'] === canvas)
+            if (match.length === 1) {
+              match[0]._rendering = true;
+              iRender.drawToTexture(match[0]['_index'], itemId, 'dupvideosource').then(res => {
+                resolve(true);
+              })
+            }
+          })
+        } else {
+          reject(Error('Invalid parameters provided. Please check if the canvas/source is correct.'));
+        }
+      })
+    }
+
+    /**
+     * params: (canvas: HTMLCanvasElement)
+     *
+     * view parameter can be 0 (Main view) or 1 (Preview).
+     *
+     */
+    static renderWorkspace(canvas: HTMLCanvasElement) {
+      return new Promise((resolve) => {
+        const match  = Render._canvasPool.filter(renObj => renObj['_canvas'] === canvas)
```

Comment:
nitpicking, double space.

```suggestion
        const match = Render._canvasPool.filter(renObj => renObj['_canvas'] === canvas)
```

It might also make sense to simply use `find` instead of filter.